## Cleaning Datasets

#### Overview Datasets:

`JA_Index_df`
 - original Judicial Autonomy Index dataset
 - includes all question items, indicators, sub-indices, and indices for Q1 and Q2

`q1_q2_df`
 - subset of Judicial Autonomy Index dataset on judges' obligations & judges' rights
 - includes indicators, sub-indices, and indices for Q1 and Q2

`q1q2_mean_df` 
 - new dataset on judges' obligations & judges' rights, merging scores for Q1 and Q2
 - includes aggregated scores (mean) of Q1 and Q2 indicators, sub-indices, and indices
 
`q1q2_oblig_df`
 - includes values for all question items on judges' obligations from Q1 and Q2
 - includes scores for the sub-index judges' obligations
 - includes scores for the Q1 and Q2 indicators

### 1 Load Dataset

In [14]:
# Load packages
import pandas as pd
import numpy as np

In [3]:
# Load dataset
JA_Index_df = pd.read_csv('../data/raw/JA_Index_v1.0.csv')
JA_Index_df

country country_code  year country_year  q1c1_apjufc  q1c1_apjuac  \
0           Albania          ALA  2000        ALA00          0.5          0.5   
1           Albania          ALA  2001        ALA01          0.5          0.5   
2           Albania          ALA  2002        ALA02          0.5          0.5   
3           Albania          ALA  2003        ALA03          0.5          0.5   
4           Albania          ALA  2004        ALA04          0.5          0.5   
..              ...          ...   ...          ...          ...          ...   
961  United Kingdom          GBR  2018        GBR18          0.5          0.5   
962  United Kingdom          GBR  2019        GBR19          0.5          0.5   
963  United Kingdom          GBR  2020        GBR20          0.5          0.5   
964  United Kingdom          GBR  2021        GBR21          0.5          0.5   
965  United Kingdom          GBR  2022        GBR22          0.5          0.5   

     q1c1_apjuhc  q1c1_appresidfc  q1c1_appresidac  q1c1_appresidhc  ...  \
0            0.0              0.5              0.5              0.0  ...   
1            0.0              0.5              0.5              0.0  ...   
2            0.0              0.5              0.5              0.0  ...   
3            0.0              0.5              0.5              0.0  ...   
4            0.0              0.5              0.5              0.0  ...   
..           ...              ...              ...              ...  ...   
961          0.5              0.5              0.5              0.5  ...   
962          0.5              0.5              0.5              0.5  ...   
963          0.5              0.5              0.5              0.5  ...   
964          0.5              0.5              0.5              0.5  ...   
965          0.5              0.5              0.5              0.5  ...   

     q2c4_case  q2c4_staff  q2_select  q2_right  q2_oblig  q2_admin  q2JA_wm  \
0        0.333         1.0      0.200     0.889     0.045     0.726    0.438   
1        0.333         1.0      0.200     0.889     0.045     0.726    0.438   
2        0.333         1.0      0.200     0.889     0.045     0.726    0.438   
3        0.333         1.0      0.200     0.889     0.113     0.726    0.448   
4        0.333         1.0      0.200     0.889     0.113     0.726    0.448   
..         ...         ...        ...       ...       ...       ...      ...   
961      0.333         1.0      0.321     0.687     0.166     0.507    0.408   
962      0.333         1.0      0.321     0.687     0.166     0.507    0.408   
963      0.333         1.0      0.321     0.687     0.166     0.507    0.408   
964      0.333         1.0      0.321     0.687     0.166     0.507    0.408   
965      0.333         1.0      0.321     0.687     0.166     0.507    0.408   

     q2JA_wp  JAI_wm  JAI_wp  
0      0.294   0.361   0.252  
1      0.294   0.361   0.252  
2      0.294   0.361   0.252  
3      0.338   0.366   0.274  
4      0.338   0.384   0.283  
..       ...     ...     ...  
961    0.374   0.372   0.304  
962    0.374   0.372   0.304  
963    0.374   0.372   0.304  
964    0.374   0.372   0.304  
965    0.374   0.372   0.304  

[966 rows x 135 columns]

### 2 Subset Dataset: Judges' Obligations & Judges' Rights

In [6]:
# Subset: filter for country-year informatio, sub-indices, indicators and index

# For Q1:
## country-year information: 'country', 'country_code', 'year', 'country_year'
## sub-indices judges rights: 'q1c2_tenure', 'q1c2_term', 'q1c2_immunity', 'q1c2_transfer', 'q1c2_finance'
## sub-indices judges obligations: 'q1c3_proced', 'q1c3_actor', 'q1c3_conflict'
## indicators: 'q1_select', 'q1_admin', 'q1_right', 'q1_oblig'
## index:'q1JA_wm', 'q1JA_wp'

# Q2:
## sub-indices judges rights: 'q2c2_term', 'q2c2_immunity', 'q2c2_removal', 'q2c2_finance'
## sub-indices judges obligations: 'q2c3_proced', 'q2c3_actor', 'q2c3_conflict'
## indicators: 'q2_select', 'q2_admin', 'q2_right', 'q2_oblig'
## index: 'q2JA_wm', 'q2JA_wp'

In [25]:
q1_q2_df = JA_Index_df[['country', 'country_code', 'year', 'country_year',
               'q1c2_tenure', 'q1c2_term', 'q1c2_immunity', 'q1c2_transfer', 
               'q1c2_finance', 'q1c3_proced', 'q1c3_actor', 'q1c3_conflict',
               'q1_select', 'q1_admin', 'q1_right', 'q1_oblig',
               'q1JA_wm', 'q1JA_wp',
               'q2c2_term', 'q2c2_immunity', 'q2c2_removal', 'q2c2_finance', 
               'q2c3_proced', 'q2c3_actor', 'q2c3_conflict',
               'q2_select', 'q2_admin', 'q2_right', 'q2_oblig',
               'q2JA_wm', 'q2JA_wp', 
               'JAI_wm', 'JAI_wp']]

q1_q2_df.to_csv('../data/clean/q1_q2_df.csv', index=False)

In [5]:
# Dimensions of dataframe: 966 rows x 33 columns
q1_q2_df.shape

(966, 33)

In [6]:
# Number of countries: 42
q1_q2_df['country'].nunique()

42

### 3 New Dataset: Merge Indicators, Sub-indices and Indices for Q1 and Q2

In [13]:
# Check original column names
q1_q2_df.columns

Index(['country', 'country_code', 'year', 'country_year', 'q1c2_tenure',
       'q1c2_term', 'q1c2_immunity', 'q1c2_transfer', 'q1c2_finance',
       'q1c3_proced', 'q1c3_actor', 'q1c3_conflict', 'q1_select', 'q1_admin',
       'q1_right', 'q1_oblig', 'q1JA_wm', 'q1JA_wp', 'q2c2_term',
       'q2c2_immunity', 'q2c2_removal', 'q2c2_finance', 'q2c3_proced',
       'q2c3_actor', 'q2c3_conflict', 'q2_select', 'q2_admin', 'q2_right',
       'q2_oblig', 'q2JA_wm', 'q2JA_wp', 'JAI_wm', 'JAI_wp'],
      dtype='str')

In [5]:
# Prepare column names for merge 

def rename_col(col):
    if 'c2' in col:
        return col.replace('c2', '_ind')
    elif 'c3' in col:
        return col.replace('c3', '_ind')
    else:
        return col
        
q1_q2_df_renamed = q1_q2_df.rename(columns=rename_col)
q1_q2_df_renamed.columns

Index(['country', 'country_code', 'year', 'country_year', 'q1_ind_tenure',
       'q1_ind_term', 'q1_ind_immunity', 'q1_ind_transfer', 'q1_ind_finance',
       'q1_ind_proced', 'q1_ind_actor', 'q1_ind_conflict', 'q1_select',
       'q1_admin', 'q1_right', 'q1_oblig', 'q1JA_wm', 'q1JA_wp', 'q2_ind_term',
       'q2_ind_immunity', 'q2_ind_removal', 'q2_ind_finance', 'q2_ind_proced',
       'q2_ind_actor', 'q2_ind_conflict', 'q2_select', 'q2_admin', 'q2_right',
       'q2_oblig', 'q2JA_wm', 'q2JA_wp', 'JAI_wm', 'JAI_wp'],
      dtype='str')

In [12]:
# Check new column names
list(q1_q2_df_renamed.columns)

['country',
 'country_code',
 'year',
 'country_year',
 'q1_ind_tenure',
 'q1_ind_term',
 'q1_ind_immunity',
 'q1_ind_transfer',
 'q1_ind_finance',
 'q1_ind_proced',
 'q1_ind_actor',
 'q1_ind_conflict',
 'q1_select',
 'q1_admin',
 'q1_right',
 'q1_oblig',
 'q1JA_wm',
 'q1JA_wp',
 'q2_ind_term',
 'q2_ind_immunity',
 'q2_ind_removal',
 'q2_ind_finance',
 'q2_ind_proced',
 'q2_ind_actor',
 'q2_ind_conflict',
 'q2_select',
 'q2_admin',
 'q2_right',
 'q2_oblig',
 'q2JA_wm',
 'q2JA_wp',
 'JAI_wm',
 'JAI_wp']

In [15]:
# Create new dataframe with mean of Q1 and Q2 indicators, sub-indices, and index

pair_map = {
    'q1q2_ind_tenure_term': ['q1_ind_tenure','q1_ind_term','q2_ind_term'],
    'q1q2_ind_immunity': ['q1_ind_immunity', 'q2_ind_immunity'],
    'q1q2_ind_removal': ['q1_ind_transfer', 'q2_ind_removal'],
    'q1q2_ind_finance': ['q1_ind_finance', 'q2_ind_finance'],
    'q1q2_ind_proced': ['q1_ind_proced', 'q2_ind_proced'],
    'q1q2_ind_actor': ['q1_ind_actor', 'q2_ind_actor'],
    'q1q2_ind_conflict': ['q1_ind_conflict', 'q2_ind_conflict'],
    'q1q2_sub-index_select': ['q1_select',  'q2_select'],
    'q1q2_sub-index_admin': ['q1_admin',  'q2_admin'],
    'q1q2_sub-index_right': ['q1_right',  'q2_right'],
    'q1q2_sub-index_oblig': ['q1_oblig',  'q2_oblig'],
    'q1q2_index_wm': ['q1JA_wm',  'q2JA_wm'],
    'q1q2_index_wp': ['q1JA_wp',  'q2JA_wp'],
}

base_columns = ['country', 'country_code', 'year', 'country_year']
q1q2_mean_df = q1_q2_df_renamed[base_columns].copy()

for new_col, cols in pair_map.items():
    q1q2_mean_df[new_col] = q1_q2_df_renamed[cols].mean(axis=1)
    
q1q2_mean_df.columns.tolist()

['country',
 'country_code',
 'year',
 'country_year',
 'q1q2_ind_tenure_term',
 'q1q2_ind_immunity',
 'q1q2_ind_removal',
 'q1q2_ind_finance',
 'q1q2_ind_proced',
 'q1q2_ind_actor',
 'q1q2_ind_conflict',
 'q1q2_sub-index_select',
 'q1q2_sub-index_admin',
 'q1q2_sub-index_right',
 'q1q2_sub-index_oblig',
 'q1q2_index_wm',
 'q1q2_index_wp']

In [41]:
q1q2_mean_df

country country_code  year country_year  q1q2_ind_tenure_term  \
0           Albania          ALA  2000        ALA00              0.666667   
1           Albania          ALA  2001        ALA01              0.666667   
2           Albania          ALA  2002        ALA02              0.666667   
3           Albania          ALA  2003        ALA03              0.666667   
4           Albania          ALA  2004        ALA04              0.666667   
..              ...          ...   ...          ...                   ...   
961  United Kingdom          GBR  2018        GBR18              0.500000   
962  United Kingdom          GBR  2019        GBR19              0.500000   
963  United Kingdom          GBR  2020        GBR20              0.500000   
964  United Kingdom          GBR  2021        GBR21              0.500000   
965  United Kingdom          GBR  2022        GBR22              0.500000   

     q1q2_ind_immunity  q1q2_ind_removal  q1q2_ind_finance  q1q2_ind_proced  \
0                 0.75              0.75               0.5           0.3125   
1                 0.75              0.75               0.5           0.3125   
2                 0.75              0.75               0.5           0.3125   
3                 0.75              0.75               0.5           0.3125   
4                 0.75              0.75               0.5           0.3125   
..                 ...               ...               ...              ...   
961               1.00              0.50               1.0           0.0000   
962               1.00              0.50               1.0           0.0000   
963               1.00              0.50               1.0           0.0000   
964               1.00              0.50               1.0           0.0000   
965               1.00              0.50               1.0           0.0000   

     q1q2_ind_actor  q1q2_ind_conflict  q1q2_sub-index_select  \
0            0.3335             0.6670                 0.2205   
1            0.3335             0.6670                 0.2205   
2            0.3335             0.6670                 0.2205   
3            0.3335             0.8335                 0.2205   
4            0.3335             1.0000                 0.2205   
..              ...                ...                    ...   
961          0.4160             0.6670                 0.3720   
962          0.4160             0.6670                 0.3720   
963          0.4160             0.6670                 0.3720   
964          0.4160             0.6670                 0.3720   
965          0.4160             0.6670                 0.3720   

     q1q2_sub-index_admin  q1q2_sub-index_right  q1q2_sub-index_oblig  \
0                   0.594                0.4545                0.1770   
1                   0.594                0.4545                0.1770   
2                   0.594                0.4545                0.1770   
3                   0.594                0.4545                0.2110   
4                   0.594                0.4545                0.3260   
..                    ...                   ...                   ...   
961                 0.457                0.4840                0.0895   
962                 0.457                0.4840                0.0895   
963                 0.457                0.4840                0.0895   
964                 0.457                0.4840                0.0895   
965                 0.457                0.4840                0.0895   

     q1q2_index_wm  q1q2_index_wp  
0           0.3610         0.2515  
1           0.3610         0.2515  
2           0.3610         0.2515  
3           0.3660         0.2735  
4           0.3835         0.2830  
..             ...            ...  
961         0.3715         0.3035  
962         0.3715         0.3035  
963         0.3715         0.3035  
964         0.3715         0.3035  
965         0.3715         0.3035  

[966 rows x 17 columns]

In [24]:
# Export dataset
q1q2_mean_df.to_csv('../data/clean/q1q2_mean_df.csv', index=False)

### 4 New Dataset: Create Subset for Judges' Obligations (incl. Questions)

In [17]:
# Filter for question items in Q1
JA_Index_df.filter(like='q1c3').columns.tolist()

['q1c3_fairtrial',
 'q1c3_disciplaw',
 'q1c3_discipints',
 'q1c3_discipbody',
 'q1c3_initdiscip',
 'q1c3_decdiscip',
 'q1c3_sanctscale',
 'q1c3_appealdiscip',
 'q1c3_judisclos',
 'q1c3_jurestrict',
 'q1c3_jurecuse',
 'q1c3_proced',
 'q1c3_actor',
 'q1c3_conflict']

In [18]:
# Filter for question items in Q2
JA_Index_df.filter(like='q2c3').columns.tolist()

['q2c3_fairtrial',
 'q2c3_disciplaw',
 'q2c3_discipints',
 'q2c3_initdiscip',
 'q2c3_decidediscip',
 'q2c3_sanctscale',
 'q2c3_judisclos',
 'q2c3_jurestrict',
 'q2c3_jurecuse',
 'q2c3_proced',
 'q2c3_actor',
 'q2c3_conflict']

In [32]:
# Subset: filter for country-year informatio and question items from judges' obligations

# For Q1: 'q1c3_fairtrial', 'q1c3_disciplaw',  'q1c3_discipints', 'q1c3_discipbody', 'q1c3_initdiscip', 'q1c3_decdiscip', 'q1c3_sanctscale', 'q1c3_appealdiscip', 'q1c3_judisclos', 'q1c3_jurestrict', 'q1c3_jurecuse'

# For Q2: 'q2c3_fairtrial', 'q2c3_disciplaw', 'q2c3_discipints', 'q2c3_initdiscip', 'q2c3_decidediscip', 'q2c3_sanctscale', 'q2c3_judisclos', 'q2c3_jurestrict', 'q2c3_jurecuse'                      

In [19]:
q1q2_oblig_items = JA_Index_df[['country', 'country_code', 'year', 'country_year',
                                'q1c3_fairtrial', 'q1c3_disciplaw',  'q1c3_discipints', 'q1c3_discipbody', 
                                'q1c3_initdiscip', 'q1c3_decdiscip', 'q1c3_sanctscale', 'q1c3_appealdiscip', 
                                'q1c3_judisclos', 'q1c3_jurestrict', 'q1c3_jurecuse',
                                'q2c3_fairtrial', 'q2c3_disciplaw', 'q2c3_discipints', 'q2c3_initdiscip', 
                                'q2c3_decidediscip', 'q2c3_sanctscale', 'q2c3_judisclos', 'q2c3_jurestrict', 
                                'q2c3_jurecuse']]
q1q2_oblig_items

country country_code  year country_year  q1c3_fairtrial  \
0           Albania          ALA  2000        ALA00               1   
1           Albania          ALA  2001        ALA01               1   
2           Albania          ALA  2002        ALA02               1   
3           Albania          ALA  2003        ALA03               1   
4           Albania          ALA  2004        ALA04               1   
..              ...          ...   ...          ...             ...   
961  United Kingdom          GBR  2018        GBR18               0   
962  United Kingdom          GBR  2019        GBR19               0   
963  United Kingdom          GBR  2020        GBR20               0   
964  United Kingdom          GBR  2021        GBR21               0   
965  United Kingdom          GBR  2022        GBR22               0   

     q1c3_disciplaw  q1c3_discipints  q1c3_discipbody  q1c3_initdiscip  \
0                 1                0                1             0.00   
1                 1                0                1             0.00   
2                 1                0                1             0.00   
3                 1                0                1             0.00   
4                 1                0                1             0.00   
..              ...              ...              ...              ...   
961               0                0                1             0.33   
962               0                0                1             0.33   
963               0                0                1             0.33   
964               0                0                1             0.33   
965               0                0                1             0.33   

     q1c3_decdiscip  ...  q1c3_jurecuse  q2c3_fairtrial  q2c3_disciplaw  \
0              0.67  ...              1             0.0             0.0   
1              0.67  ...              1             0.0             0.0   
2              0.67  ...              1             0.0             0.0   
3              0.67  ...              1             0.0             0.0   
4              0.67  ...              1             0.0             0.0   
..              ...  ...            ...             ...             ...   
961            0.00  ...              1             0.0             0.0   
962            0.00  ...              1             0.0             0.0   
963            0.00  ...              1             0.0             0.0   
964            0.00  ...              1             0.0             0.0   
965            0.00  ...              1             0.0             0.0   

     q2c3_discipints  q2c3_initdiscip  q2c3_decidediscip  q2c3_sanctscale  \
0                0.0              0.0                0.0              0.0   
1                0.0              0.0                0.0              0.0   
2                0.0              0.0                0.0              0.0   
3                0.0              0.0                0.0              0.0   
4                0.0              0.0                0.0              0.0   
..               ...              ...                ...              ...   
961              0.0              1.0                0.0              0.0   
962              0.0              1.0                0.0              0.0   
963              0.0              1.0                0.0              0.0   
964              0.0              1.0                0.0              0.0   
965              0.0              1.0                0.0              0.0   

     q2c3_judisclos  q2c3_jurestrict  q2c3_jurecuse  
0               0.0              1.0            1.0  
1               0.0              1.0            1.0  
2               0.0              1.0            1.0  
3               1.0              1.0            1.0  
4               1.0              1.0            1.0  
..              ...              ...            ...  
961             0.0              1.0            1.0  
962         

In [21]:
# Merge with sub-index judges obligations and indicators: 'q1q2_sub-index_oblig', 'q1q2_ind_proced', 'q1q2_ind_actor', 'q1q2_ind_conflict'

merge_df = q1q2_mean_df[['country_year', 'q1q2_sub-index_oblig', 'q1q2_ind_proced', 'q1q2_ind_actor', 'q1q2_ind_conflict']]
merge_df

q1q2_oblig_df = pd.merge(q1q2_oblig_items, merge_df, how='outer', on='country_year')
q1q2_oblig_df

country country_code  year country_year  q1c3_fairtrial  q1c3_disciplaw  \
0    Albania          ALA  2000        ALA00               1               1   
1    Albania          ALA  2001        ALA01               1               1   
2    Albania          ALA  2002        ALA02               1               1   
3    Albania          ALA  2003        ALA03               1               1   
4    Albania          ALA  2004        ALA04               1               1   
..       ...          ...   ...          ...             ...             ...   
961  Ukraine          UKR  2018        UKR18               1               1   
962  Ukraine          UKR  2019        UKR19               1               1   
963  Ukraine          UKR  2020        UKR20               1               1   
964  Ukraine          UKR  2021        UKR21               1               1   
965  Ukraine          UKR  2022        UKR22               1               1   

     q1c3_discipints  q1c3_discipbody  q1c3_initdiscip  q1c3_decdiscip  ...  \
0                  0                1              0.0            0.67  ...   
1                  0                1              0.0            0.67  ...   
2                  0                1              0.0            0.67  ...   
3                  0                1              0.0            0.67  ...   
4                  0                1              0.0            0.67  ...   
..               ...              ...              ...             ...  ...   
961                1                1              1.0            1.00  ...   
962                1                1              1.0            1.00  ...   
963                1                1              1.0            1.00  ...   
964                1                1              1.0            1.00  ...   
965                1                1              1.0            1.00  ...   

     q2c3_initdiscip  q2c3_decidediscip  q2c3_sanctscale  q2c3_judisclos  \
0                0.0                0.0              0.0             0.0   
1                0.0                0.0              0.0             0.0   
2                0.0                0.0              0.0             0.0   
3                0.0                0.0              0.0             1.0   
4                0.0                0.0              0.0             1.0   
..               ...                ...              ...             ...   
961              1.0                1.0              1.0             1.0   
962              1.0                1.0              1.0             1.0   
963              1.0                1.0              1.0             1.0   
964              1.0                1.0              1.0             1.0   
965              1.0                1.0              1.0             1.0   

     q2c3_jurestrict  q2c3_jurecuse  q1q2_sub-index_oblig  q1q2_ind_proced  \
0                1.0            1.0                0.1770           0.3125   
1                1.0            1.0                0.1770           0.3125   
2                1.0            1.0                0.1770           0.3125   
3                1.0            1.0                0.2110           0.3125   
4                1.0            1.0                0.3260           0.3125   
..               ...            ...                   ...              ...   
961              1.0            1.0                0.9345           1.0000   
962              1.0            1.0                0.9345           1.0000   
963              1.0            1.0                0.9345           1.0000   
964              1.0            1.0                0.9345           1.0000   
965              1.0            1.0                0.9345           1.0000   

     q1q2_ind_actor  q1q2_ind_conflict  
0            0.3335             0.6670  
1            0.3335             0.6670  
2            0.3335             0.6670  
3            0.3335             0.8335  
4            0.3335             1.0000  

In [23]:
q1q2_oblig_df.to_csv('../data/clean/q1q2_oblig_df.csv', index=False)

### 5 Check Missing Values

In [11]:
# Missing values
q1_q2_df.isna().sum()

country            0
country_code       0
year               0
country_year       0
q1c2_tenure        0
q1c2_term          0
q1c2_immunity      0
q1c2_transfer      0
q1c2_finance       0
q1c3_proced        0
q1c3_actor         0
q1c3_conflict      0
q1_select          0
q1_admin           0
q1_right           0
q1_oblig           0
q1JA_wm            0
q1JA_wp            0
q2c2_term        115
q2c2_immunity    115
q2c2_removal     116
q2c2_finance     115
q2c3_proced      203
q2c3_actor       260
q2c3_conflict    115
q2_select        115
q2_admin         115
q2_right         115
q2_oblig         115
q2JA_wm          115
q2JA_wp          115
JAI_wm             0
JAI_wp             0
dtype: int64

In [12]:
# Create overview of country-years with missing values per column (V1)

def overview_nan_df(df):
    
    # Create empty dictionary
    nan_dict = {}
    
    # Iterate through all columns of dataframe:
    for col in df.columns:
        # if column has missing values, create list of country_year where that missing value occurs
        if df[col].isna().any():
            missing_country_years = df.loc[df[col].isna(), 'country_year'].tolist()
           # missing_country_years = ", ".join(map(str, missing_country_years))
            nan_dict[col] = missing_country_years

    return nan_dict

In [13]:
df_missing_values = overview_nan_df(q1_q2_df)
df_missing_values

{'q2c2_term': ['FIN00',
  'FIN01',
  'FIN02',
  'FIN03',
  'FIN04',
  'FIN05',
  'FIN06',
  'FIN07',
  'FIN08',
  'FIN09',
  'FIN10',
  'FIN11',
  'FIN12',
  'FIN13',
  'FIN14',
  'FIN15',
  'FIN16',
  'FIN17',
  'FIN18',
  'FIN19',
  'FIN20',
  'FIN21',
  'FIN22',
  'ISL00',
  'ISL01',
  'ISL02',
  'ISL03',
  'ISL04',
  'ISL05',
  'ISL06',
  'ISL07',
  'ISL08',
  'ISL09',
  'ISL10',
  'ISL11',
  'ISL12',
  'ISL13',
  'ISL14',
  'ISL15',
  'ISL16',
  'ISL17',
  'ISL18',
  'ISL19',
  'ISL20',
  'ISL21',
  'ISL22',
  'NLD00',
  'NLD01',
  'NLD02',
  'NLD03',
  'NLD04',
  'NLD05',
  'NLD06',
  'NLD07',
  'NLD08',
  'NLD09',
  'NLD10',
  'NLD11',
  'NLD12',
  'NLD13',
  'NLD14',
  'NLD15',
  'NLD16',
  'NLD17',
  'NLD18',
  'NLD19',
  'NLD20',
  'NLD21',
  'NLD22',
  'NOR00',
  'NOR01',
  'NOR02',
  'NOR03',
  'NOR04',
  'NOR05',
  'NOR06',
  'NOR07',
  'NOR08',
  'NOR09',
  'NOR10',
  'NOR11',
  'NOR12',
  'NOR13',
  'NOR14',
  'NOR15',
  'NOR16',
  'NOR17',
  'NOR18',
  'NOR19',
  'NOR20

In [14]:
# Create overview of country-years with missing values per column (V2 -- Advanced)

def overview_nan_dict_adv(df):
    
    result = {}
    
    for col in df.columns:
        if df[col].isna().any():
            
            missing = df.loc[df[col].isna(), "country_year"]
            
            temp = (pd.DataFrame({"country_year": missing}))
            
            temp["country"] = temp["country_year"].str[:3]
            temp["year"] = temp["country_year"].str[3:].astype(int) + 2000
            
            nested = (
                temp.groupby("country")["year"]
                .apply(list)
                .to_dict()
            )
            
            result[col] = nested
    
    return result

In [15]:
# Return overview of country-years with missing values (V1)

df_missing_values_adv = overview_nan_dict_adv(q1_q2_df)

for outer_key, inner_dict in df_missing_values_adv.items():
    print(f"\n*** Question item:{outer_key} ***")
    for inner_key, inner_value in inner_dict.items():
        print(f"  {inner_key}: {inner_value}")


*** Question item:q2c2_term ***
  CHE: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]
  FIN: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]
  ISL: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]
  NLD: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]
  NOR: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]

*** Question item:q2c2_immunity ***
  CHE: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]
  FIN: [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007

In [16]:
# Return overview of country-years with missing values (V2)

df_missing_values_adv = overview_nan_dict_adv(q1_q2_df)

for key, value in df_missing_values_adv.items():
    print(f"\nQuestion item: {key}")
    print(f"Country years: {value}")


Question item: q2c2_term
Country years: {'CHE': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022], 'FIN': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022], 'ISL': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022], 'NLD': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022], 'NOR': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022]}

Question item: q2c2_immunity
Country years: {'CHE': [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022], 'FIN': [2000, 2001, 2002, 2003

#### SUMMARY

##### q2c2_term (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2c2_immunity (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2c2_removal (missing values: 116)      
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022, Montenegro    2022

##### q2c2_finance (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2c3_proced (missing values: 203) 
- Armenia 2000-2018, Azerbaijan 2000-2022, Finland 2000-2022, Island 2000-2022, Montenegro 2000-2022, Netherlands    2000-2022, Norway 2000-2022, Slovenia 2000-2022, Switzerland 2000-2022

##### q2c3_actor (missing values: 260)   
- Armenia 2000-2018, Azerbaijan 2000-2022, Bulgaria 2000-2022, Croatia 2000-2022, Finland 2000-2022, Island 2000-    2022, Montenegro 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Slovenia 2000-2022, Switzerland 2000-2022,    Turkey 2000-2010

##### q2c3_conflict (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2_select (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2_admin (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2_right (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2_oblig (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2JA_wm (missing values: 115) 
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022

##### q2JA_wp (missing values: 115)
- Finland 2000-2022, Island 2000-2022, Netherlands 2000-2022, Norway 2000-2022, Switzerland 2000-2022